tutorials/reddit_slack_bot/reddit_slack_bot.py

```diff
@@ -0,0 +1,169 @@
+# time we ran the workflow; specified by `DAYS_BETWEEN_RUNS`.
+
+# We also make use of caching by setting `cache=True` in the `@task` decorator. By adding
+# `kickoff_time` and `lookback_days` as arguments to the task we keep the function code
+# independent of the current time. Now if `get_posts` ever gets called multiple times, we will
+# not make unnecessary requets to the reddit API.
+
+@task(
+    secret_requests=[
+        Secret(key="reddit_client_id"),
+        Secret(key="reddit_secret_key"),
+    ],
+    cache=True,
+)
+def get_posts(kickoff_time: datetime, lookback_days: int, search_terms: List[str]) -> List[Dict[str, str]]:
+    # Load Secrets and Authenticate
+    reddit_client_id = current_context().secrets.get("reddit_client_id")
+    reddit_secret_key = current_context().secrets.get("reddit_secret_key")
+    auth = HTTPBasicAuth(reddit_client_id, reddit_secret_key)
+
+    # Format and run request for reddit posts
+    query_string = '+'.join(search_terms)
+    result = requests.get(f'https://www.reddit.com/search.json?q={query_string}&sort=new', auth=auth)
+    result_json = result.json()
+    posts = result_json['data']['children']
+
+    # Get only recent posts
+    days_ago_datetime = kickoff_time - timedelta(days=lookback_days)
+    days_ago_timestamp = days_ago_datetime.timestamp()
+    recent_posts = [{
+        'title': post['data']['title'],
+        'description': post['data']['selftext'],
+        'link': post['data']['url']
+    } for post in posts if post['data']['created_utc'] >= days_ago_timestamp]
+
+    return recent_posts
+
+# ## Posting to Slack
+#
+# Given the reddit posts returned by our previous task, we define a helper function to format them
+# in a more readable format for Slack.
+
+def format_posts(posts: List[Dict[str, str]]):
+    formatted_message = "*Recent Posts:*\n\n"
+    for idx, post in enumerate(posts, 1):
+        formatted_message += f"*{idx}. <{post['link']}|{post['title']}>*\n"
+        formatted_message += f"{post['description']}\n\n"
+    return formatted_message
+
+# We define another task to send our posts to slack. This time we use the `ImageSpec`
+# we previously wrote that contains the `slack_sdk` package.
+
+@task(
+    container_image=image,
+    secret_requests=[Secret(key="slack_token")]
+)
+def post_slack_message(recent_posts: List[Dict[str, str]]):
+    from slack_sdk import WebClient
+
+    slack_token = current_context().secrets.get("slack_token")
+    client = WebClient(token=slack_token)
+
+    response = client.chat_postMessage(
+        channel=SLACK_CHANNEL_NAME,
+        text=format_posts(recent_posts)
+    )
+    assert response["message"]["text"]
+
+# ## Creating the Workflow
+#
+# Finally, we chain these tasks together into a simple two-step workflow with some default inputs.
+
+@workflow
+def reddit_wf(kickoff_time: datetime = datetime(2024, 1, 1), lookback_days: int = DAYS_BETWEEN_RUNS, search_terms: List[str]= ["flyte", "ml"]):
```

Comment:
Can you add docstrings like this in all tasks and workflows?
```suggestion
def reddit_wf(kickoff_time: datetime = datetime(2024, 1, 1), lookback_days: int = DAYS_BETWEEN_RUNS, search_terms: List[str]= ["flyte", "ml"]):
    """
    <Add a description here>
    
    :param kickoff_time: when was this execution kicked of
    ...
```